### ISSUE TITLE
Keep the terminal history / keep terminal running

### ISSUE THREAD

**Issue by hexa00** [theia-ide/theia#821]
As I user I would like to be able to:
 
   - Close my window and leave a terminal running and see the history and output on workspace restore
  -  Open multiple windows on the same workspace and be able to share the terminals output ?
   - See the output of a exited task

**Comment by hexa00** [comment]
User stories about the Terminal:

  - User starts a terminal and closes it:  Terminal should close in the
    backend.
 
  - User starts a terminal and closes its browser: Terminal should stay
    open in the backend.

  - User restores a workspace with active terminals: Terminal should
    reconnect to active terminals and display history.

  - User shares its workspace, both users should see the same terminal and
    workspace...? (This is not in scope for this PR but the code should
    work with that assuming the workspace restoration works properly)

  - Maximum history should be configurable: default max 10MB ?


User stories for tasks:

  - User runs a task and the tasks exists: User should still see the
    history.

  - User restores a workspace with exited tasks: User should still see their
    history.

  - User clears tasks history by using the Clear Exited Tasks command.

  - Could have an option: restore exited tasks: true/false if someone find
    this clutters it's workspace.


Electron specific considerations:

For  electron running a local backend the background tasks will be
closed when the application closes. So failing to restore a terminal
should be normal and the terminal should just not show.

**Comment by hexa00** [comment]
Design discussion:


The first idea that comes to mind to implement this is to have a class
like MultiOuputBuffer that would look like (note that is meant as pseudocode):

```typescript
class MultiOuputBuffer 

 buffer: RingBuffer[MAX_SIZE] 
 bufferPassThrough: PassThroughStream;

 constructor (input: ReadalbeStream) {
  input.pipe(bufferPassThrough)
  bufferPassTrhough.on('data', data => buffer.push(data))
 }

/* This will pause the input stream
   Send the buffered history
   Resume the input stream,
   (To keep the history consistent)
*/
 pipe(dest: ReadableStream) {
   input.pause(); 
   sendBufferedData(dest);
   input.pipe(dest);
   input.resume();
 }  

```

Using this method has the following problems:

  - The backend sends potentially a lot of data to the frontend
  - While that data is sent the output other terminals will be frozen
 
 
Possible solutions to these problems:

 - Store the history in the frontend:

 While this would avoid sending a lot of data from the backend to the
 frontend it would have the following caveats:

  - Terminal history is not sharable between frontends
  - If there's a detached terminal, when the session is restored
    the history will be missing the data since last session.

I think these two problems disqualify this solution.

 - Use a pull based data fetching from the frontend:

The idea would be that the frontend would request data from the backend
in blocks so that scrolling remains smooth but not all data is fetched at
one time.

As the user scrolls back more data would be fetched from the backend to
show the user.

  This has the following problems however:

   - The xterm.js terminal however is not meant for that. It controls
its scrolling etc and usually works by appending data only. It may need
some work to change xterm.js to work like this.

   - It will most likely break ncurse based applications since they may
require more previous data than available to work properly.


So this solution is disqualified too.


 - Send the terminal history asynchronously

The idea would be to reduce the time that the other terminal are frozen by
sending the buffered history without blocking the source.

And once that is done send what has accumulated in the buffer while we
were sending the data.

Something like:

```typescript

function pipe(dest) {

/* Note we need to copy the buffer in memory here.. since we keep
   appending to it... but there may be better ways, i'm not sure yet
   how the buffered data will be structured we just need to keep a stable
   point to send data from */
   sendBufferedDataAsync(dest, buffer.slice(0), then(() => {
     input.pause()
     sendData(tempBuffer);
     input.pipe(dest); 
     input.unpipe(tmpStream);
     input.resume()
   } 

   tmpBuffer: Buffer[]
   tmpStream.on('data', data => tmpBuffer.push(data))
   input.pipe(tmpStream);
 }  

```

This way will still send possibly a lot of a data but the terminal should
not freeze much.

There may be issues based on the rate at which output is produce for example if the network between the frontend / backend is slower than the data produced by the terminal. We may still send a lot of data synchronously then.

I think I will go with that solution and we can further optimize it later.

Ideas are welcome

**Comment by hexa00** [comment]
Thinking more about this and talking with @simark we could make it so that we do not need the tmpBuffer thing and implement a kind of RingBufferReadStream that supports connecting multiple clients and remembers their positions.

So the clients would just connect to that and start reading from the start and record their position as the data is consumed.

I'll implement that.

**Comment by svenefftinge** [comment]
Thanks for writing this up.

__For the 'backend is blocked'-issue__ I think there will be very rare cases, where multiple users look at the same workspace including a terminal with a super long history. So the simple solution of just sending the history in chunks leaving ticks for other things in between sounds good to me. And as you said, if it turns out to be a real issue, we can go back an improve.

__For the user stories__ could we treat a terminal just like a task executing a bash? And then when closing the terminal widget connected to a TerminalProcess, I think the session should still run, so the user needs to explicitly terminate tasks. 
We will need a widget that lists the processes ordered by least recently started and active state which allows the user to open a terminal and restart/stop them.

**Comment by hexa00** [comment]
>For the 'backend is blocked'-issue
I'm going to do it non blocking with the sharing in mind from the start, it should not take too long to do.

>For the user stories
It is the plan to treat the terminal a a task, even if a special internal one.

However for now to close the widget will close the terminal as we don't have a widget like the planned task widget.

Once we have this task widget I think we could make another based on it for the terminals only. Then we can modify how they are closed. Most likey there will be a command to close a terminal inside that widget.

Sounds good ?
